api/grpcserver/v2alpha1/activation.go

```diff
@@ -0,0 +1,348 @@
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	spacemeshv2alpha1 "github.com/spacemeshos/api/release/go/spacemesh/v2alpha1"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	"github.com/spacemeshos/go-spacemesh/api/grpcserver"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/events"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+	"github.com/spacemeshos/go-spacemesh/sql/builder"
+)
+
+const (
+	Activation       = "activation_v2alpha1"
+	ActivationStream = "activation_stream_v2alpha1"
+)
+
+func NewActivationStreamService(db *sql.Database) *ActivationStreamService {
+	return &ActivationStreamService{db: db}
+}
+
+type ActivationStreamService struct {
+	db *sql.Database
+}
+
+var _ grpcserver.ServiceAPI = (*ActivationStreamService)(nil)
+
+func (s *ActivationStreamService) RegisterService(server *grpc.Server) {
+	spacemeshv2alpha1.RegisterActivationStreamServiceServer(server, s)
+}
+
+func (s *ActivationStreamService) RegisterHandlerService(mux *runtime.ServeMux) error {
+	return spacemeshv2alpha1.RegisterActivationStreamServiceHandlerServer(context.Background(), mux, s)
+}
+
+func (s *ActivationStreamService) String() string {
+	return "ActivationStreamService"
+}
+
+func (s *ActivationStreamService) Stream(
+	request *spacemeshv2alpha1.ActivationStreamRequest,
+	stream spacemeshv2alpha1.ActivationStreamService_StreamServer,
+) error {
+	var sub *events.BufferedSubscription[events.ActivationTx]
+	if request.Watch {
+		matcher := resultsMatcher{request, stream.Context()}
+		var err error
+		sub, err = events.SubscribeMatched(matcher.match)
+		if err != nil {
+			return status.Error(codes.Internal, err.Error())
+		}
+		defer sub.Close()
+		if err := stream.SendHeader(metadata.MD{}); err != nil {
+			return status.Errorf(codes.Unavailable, "can't send header")
+		}
+	}
+	ops, err := toOperations(toRequest(request))
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+	var ierr error
+	if err := atxs.IterateAtxsOps(s.db, ops, func(atx *types.VerifiedActivationTx) bool {
+		ierr = stream.Send(&spacemeshv2alpha1.Activation{Versioned: &spacemeshv2alpha1.Activation_V1{V1: toAtx(atx)}})
+		return ierr == nil
+	}); err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	if sub == nil {
+		return nil
+	}
+	for {
+		select {
+		case <-stream.Context().Done():
+			return nil
+		case <-sub.Full():
+			return status.Error(codes.Canceled, "buffer overflow")
+		case rst := <-sub.Out():
+			if err := stream.Send(&spacemeshv2alpha1.Activation{
+				Versioned: &spacemeshv2alpha1.Activation_V1{V1: toAtx(rst.VerifiedActivationTx)},
+			},
+			); err != nil {
+				if errors.Is(err, io.EOF) {
+					return nil
+				}
+				return status.Error(codes.Internal, err.Error())
+			}
```

Comment:
How about a switch that avoids multiple if nesting:
```suggestion
			err := stream.Send(&spacemeshv2alpha1.Activation{
				Versioned: &spacemeshv2alpha1.Activation_V1{V1: toAtx(rst.VerifiedActivationTx)},
			})
			switch {
			case errors.Is(err, io.EOF):
				return nil
			case err != nil:
				return status.Error(codes.Internal, err.Error())
			}
```